### PR TITLE
Debug dump utility HLD

### PR DESCRIPTION
**Pull request by vivekreddynv**
Signed-off-by: Vivek Reddy Karri vkarri@nvidia.com

The high-level design document for the dump cli utility

| Repo  | PR title | State |  
| ---------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------- | 
| sonic-utilities | [[debug dump util] Match Infrastructure](https://github.com/Azure/sonic-utilities/pull/1666) | GitHub issue/pull request detail |  
| sonic-utilities | [[debug dump util] Module implementation Logic and Port Module](https://github.com/Azure/sonic-utilities/pull/1667)  | GitHub issue/pull request detail |   
| sonic-utilities | [[debug dump util] Base Skeleton and Click Class added](https://github.com/Azure/sonic-utilities/pull/1668) | GitHub issue/pull request detail |  
| sonic-utilities | [[debug dump util] Techsupport addition](https://github.com/Azure/sonic-utilities/pull/1669) | GitHub issue/pull request detail | 
| sonic-utilities | [[debug dump util] COPP Module Added](https://github.com/Azure/sonic-utilities/pull/1670) | GitHub issue/pull request detail |

### CODE REVIEW

**Comment by gechiang** [comment]
@SuvarnaMeenakshi Can you help clarify if VS for multi-asic support is available so that @vivekreddynv can add the support of this utilities for multi-asic? 
Thanks!

**Comment by SuvarnaMeenakshi** [comment]
> @SuvarnaMeenakshi Can you help clarify if VS for multi-asic support is available so that @vivekreddynv can add the support of this utilities for multi-asic?
> Thanks!
for master branch, this PR is required to bring up multi-asic vs: https://github.com/Azure/sonic-buildimage/pull/7632
PR verification steps have information on how to bring up a multi-asic vs testbed.
Also, swss docker does not come up currently, I am debugging that part currently.
To workaround, this diff can be used for now: https://github.com/SuvarnaMeenakshi/sonic-buildimage/commit/f7f626049d06356708ce6d21012b1cdffd8255cb

**Comment by vivekreddynv** [comment]
@gechiang I've added "--namespace / -n" option to the `dump state` CLI. I'm using `SonicV2Connector` to connect to the Redis Underneath and i've made sure that this argument is propagated down to the connector level. I've also added the Multi-ASIC UT's in the sonic-utilities directory and made sure the change is working as expected. 

However, i couldn't test this feature on actual hardware or on a VS Image currently. I'll try my best to test this on a VS image once the issues mentioned by @SuvarnaMeenakshi are merged and also when i find some buffer. 

If you wish to verify the functionality on a multi-asic platform, i can send the code tarball and instructions or a binary built with these changes to test the feature

**Comment by liat-grozovik** [comment]
@lguohan can you please assign a reviewer to approve this HLD?
PRs should come by EOW and HLD should be closed before it. Thanks.

File: doc/Dump-Utility.md [hunk 1]
```diff
@@ -0,0 +1,823 @@
...
+
+    def get_all_args(self, namespace):
+        all_port_names = []
+        .......  # Find and fill this list of all_port_names,
+        return (all_ports) #Eg: (['Ethernet0', 'Ethernet4', 'Ethernet8', 'Ethernet12', ....])
+
+    def execute(self, param):
+        self.template = display_template(dbs=["CONFIG_DB", "APPL_DB", "ASIC_DB"])
+        port = param[ARG_NAME]
+	ns = param["namespace"]  # namespace info, if any provided by the user
```
Comment (SuvarnaMeenakshi) [inline doc/Dump-Utility.md:627]: In multi-asic, there could be some modules/features specific to default namespace. Ex: SNMP.
If we try to get SNMP data from CONFIG_DB, and pass a namespace, the key will not available in that namespace. so should it error out saying  table/key not found in the namespace provided?

**Review by SuvarnaMeenakshi — CHANGES_REQUESTED** [review]

**Comment by SuvarnaMeenakshi** [comment]
> @gechiang I've added "--namespace / -n" option to the `dump state` CLI. I'm using `SonicV2Connector` to connect to the Redis Underneath and i've made sure that this argument is propagated down to the connector level. I've also added the Multi-ASIC UT's in the sonic-utilities directory and made sure the change is working as expected.
> 
> However, i couldn't test this feature on actual hardware or on a VS Image currently. I'll try my best to test this on a VS image once the issues mentioned by @SuvarnaMeenakshi are merged and also when i find some buffer.
> 
> If you wish to verify the functionality on a multi-asic platform, i can send the code tarball and instructions or a binary built with these changes to test the feature

You could use the latest master branch to bring up multi-asic Virtual swith, the required changes are in place.
https://github.com/Azure/sonic-buildimage/blob/master/platform/vs/README.vsvm.md - Reference.

File: doc/Dump-Utility.md [hunk 2]
```diff
@@ -0,0 +1,823 @@
...
+                    "PORT|Ethernet0": {
+                        "admin_status": "up",
+                        "alias": "etp1",
+                        "index": "1",
+                        "lanes": "0,1,2,3,4,5,6,7",
+                        "speed": "100000"
+                    }
+                }
+            ],
+            "tables_not_found": []
```
[the remark this answers is not in the document]
Comment (vivekreddynv) [reply doc/Dump-Utility.md:126]: That column will have those table names, which are supposed to have the info which the module was looking for and yet it did not find one.

File: doc/Dump-Utility.md [hunk 3]
```diff
@@ -0,0 +1,823 @@
...
+                "oid:0xd00000000056d": "oid:0xd",
+                "oid:0x10000000004a4": "oid:0x1690000000001"
+            }
+        }
+    },
+    <Truncated>
+}
+
+# Specify the namespace from which the keys should be fetched from using --namespace
+root@sonic# dump state port Ethernet0 --db APPL_DB --namespace asic0
```
[the remark this answers is not in the document]
Comment (vivekreddynv) [reply doc/Dump-Utility.md:417]: If nothing is specified, the default namespace is used.  Yeah, unless you filter them based on --db name and if the module is programmed to look for a that, that DB name will be present in the dump output. 

So, the output will always be specific to a namespace.

File: doc/Dump-Utility.md [hunk 4]
```diff
@@ -0,0 +1,823 @@
...
+{
+  "<DB_NAME>": {
+        "keys": [],              # Relevant keys i.e. "Table<sep>key" info in the corresponding DB which are of interest
+        "tables_not_found": []   # Populate this list with those table names, which are supposed to have the info which the module was looking for and yet it did not find one.
+                                 # Again, the decision of which tables to look for and which of those are definitely supposed to have the information,
+                                   is left to the discretion of the module implementer.
+   }
+}
+Note: DB_NAME could be either of ["CONFIG_DB", "APPL_DB", "ASIC_DB", "STATE_DB", ....] or "CONFIG_FILE".
+COPP is a example which will have a entry "CONFIG_FILE". This entry is required for COPP because, the default COPP entries are read from copp_cfg.json by coppmgr and are not present in the CONFIG_DB.
```
[the remark this answers is not in the document]
Comment (vivekreddynv) [reply doc/Dump-Utility.md:574]: Currently, only JSON file supported.

File: doc/Dump-Utility.md (hunk 1 above)
Comment (vivekreddynv) [reply doc/Dump-Utility.md:627]: Yeah, that will likely happen, if not handled inside the module. 

If a module was to be programmed for SNMP and if we know, the data for SNMP will not be present in namespaces other than DEFAULT, then module implementor should take that into consideration when using MatchInfra to pull the data.

File: doc/Dump-Utility.md [hunk 5]
```diff
@@ -0,0 +1,823 @@
...
+2) A Second Argument should also be defined for a feature.
+3) This argument could either be a table-key or a unique-field-value present in either Conf DB or Appl DB.
+   * Eg: For PORT, the second argument will be an interface name i.e 'Ethernet128' which is a table-key. On the other hand, the secondary argument for COPP will be a trap_id such as  'arp_req', which is a field-value and not a key of any table.
+4) The decision of what to pass as a secondary argument lies with the discretion of the one who is writing the module.
+6) The Command should also take a list of comma seperated inputs for the secondary argument
+7) The Command should also accept an "all" value and which means it should print the unified view for every entry related to that feature.
+
+```
+root@sonic# dump state --help
+Usage: main.py state [OPTIONS] MODULE IDENTIFIER
```
[the remark this answers is not in the document]
Comment (vivekreddynv) [reply doc/Dump-Utility.md]: Resolved!

File: doc/Dump-Utility.md [hunk 6]
```diff
@@ -0,0 +1,823 @@
...
+The task of debugging quickly becomes tedious because currently, there is no utility which does print a unified view of the redis-state.
+This is the problem which is addressed by this dump utility.
+
+This utility provides the base infrastructure and guidelines to make is easy for the developers to extend and support the utility for different modules.
+
+## **1 Requirements**
+
+### 1.1 Intended Usage
+
+Note: Currently, `state` is the only subcommand under `dump`.
```
Comment (SuvarnaMeenakshi) [inline doc/Dump-Utility.md:57]: Will 'state' be confusing to understand.
For example, if we say dump "state", will use expect to dump the "state_db"? 
Will user think that, dump could either dump state or dump config.
What is the intention of adding sub command under dump? Is the idea to use dump for any thing else in future? is there an example where dump could be used for something else in future?

Comment (vivekreddynv) [reply doc/Dump-Utility.md:57]: Yeah, that was indeed the idea, if we choose to use the dump command for something else apart from fetching just redis-state. For example, Maybe to fetch data from sdk dumps, etc.

As for the state subcommand,  the help should make it clear.

```
root@sonic# dump state --help
Usage: dump state [OPTIONS] MODULE IDENTIFIER

  Dump the redis-state of the identifier for the module specified

Options:
  -s, --show            Display Modules Available
  -d, --db TEXT         Only dump from these Databases
  -t, --table           Print in tabular format  [default: False]
  -k, --key-map         Only fetch the keys matched, don't extract field-value dumps  [default: False]
  -v, --verbose         Prints any intermediate output to stdout useful for dev & troubleshooting  [default: False]
  -n, --namespace TEXT  Dump the redis-state for this namespace.  [default: DEFAULT_NAMESPACE]
  --help                Show this message and exit.
```

Comment (SuvarnaMeenakshi) [reply doc/Dump-Utility.md:57]: Ok, The help sentence should be modified to convey that the CLI is not just to dump redis-state but also can dump from a CONFIG_FILE.
Also, is there any help in the command that tells the user that -d/--db option can take db name or CONFIG_FILE?

Comment (vivekreddynv) [reply doc/Dump-Utility.md:57]: 1) Yeah, will add that
2) Currently, there is no way to notify the user about the DB/CONFIG_FILE's that the particular module will be looking, directly from the CLI. Can add this to the roadmap for next revision though.

Comment (SuvarnaMeenakshi) [reply doc/Dump-Utility.md:57]: Ok, It is better to update the help message for --db sub command, telling "Dump from specified database or config_file".

Comment (vivekreddynv) [reply doc/Dump-Utility.md:57]: Updated them in the HLD.

Comment (SuvarnaMeenakshi) [reply doc/Dump-Utility.md:57]: Help message of "dump state" could be shortened as:
 "Dump the current state of the identifier for the specified module from Redis DB or CONFIG_FILE"

Comment (vivekreddynv) [reply doc/Dump-Utility.md:57]: Done

**Comment by liat-grozovik** [comment]
@SuvarnaMeenakshi kindly reminder to review recent updates

**Review by SuvarnaMeenakshi — APPROVED** [review]
LGTM

**Pull request merged**